### PR TITLE
import DNSSEC conformance test suite repository

### PR DESCRIPTION
**Pull request by japaric**
this PR imports the git history of the dnssec-tests repository (RFC #2155) using the following command
``` console
git subtree add -P conformance https://github.com/ferrous-systems/dnssec-tests main
```

and then proceeds to set up CI and CODEOWNERS as proposed in [this comment](https://github.com/hickory-dns/hickory-dns/pull/2155#issuecomment-2083288096)

the first ~159 commits are the git history of the dnssec-tests repository. the last few commits correspond to the integration work

I considered adding the new two packages to the existing workspace but that resulted in more invasive changes to the `justfile`: we don't want to run the conformance test suite whenever `cargo ws exec cargo t` is used because we want to set the test subject via an environment variable and we also want to run the conformance test suite against different DNS implementations. I didn't find a straightforward way to merge the `ignore` justfile arguments that some tasks have to a default `--ignore='{dns-test,conformance-tests}'` argument which had to be added to each `cargo ws exec` invocation. Instead, I opted for placing the conformance test suite and its test framework into its own workspace; that led to less invasive changes to the `justfile`.

Note that the creation of the CODEOWNERS file is not sufficient to achieve the desired review permissions. The ferrous team need to be given write permissions to this repository and the branch protection settings of this repository need to modified to "Require review from Code Owners" as described in [this help article](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#codeowners-and-branch-protection)

~~I'm going to leave this PR in draft state because I need to redo the "subtree pull" after https://github.com/ferrous-systems/dnssec-tests/pull/67 has been merged because that un-ignores the tests that pass with the latest version of hickory-dns ( 107635c6c5934524894736f1b141198d0fa62fec )~~ DONE

### CODE REVIEW

File: .github/CODEOWNERS
Comment (japaric) [inline .github/CODEOWNERS:2]: this file can be simplified if we use teams instead of usernames but those teams would need to be created within the hickory-dns organization

File: .github/workflows/conformance.yml
Comment (japaric) [inline .github/workflows/conformance.yml:12]: in the future, when the conformance test suite gets larger, we may want to split this job into several. the job parallelism may reduce the wall clock time of the entire CI pipeline

File: .github/workflows/test.yml
Comment (japaric) [inline .github/workflows/test.yml:9]: according to [GitHub docs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-excluding-paths), this means "run this workflow only if a file *outside* the `conformance` directory has been changed" which is the behavior we want for this workflow.

we want to run the `conformance` workflow on its own when all the changes are inside the `conformance` directory and we want to run *both* workflows when something changes outside the `conformance` directory

**Comment by japaric** [comment]
this PR is ready for review. the CI jobs are green outside the known nightly failure reported in #2223 

I think the `conformance` workflow should also be labeled as "required" but that needs to be done in the repository settings, I presume -- I do not have permissions to change the repository settings.

2024-05-29T19:17:05+02:00

**Comment by pvdrz** [comment]
I merged the latest commits from main to fix the nightly CI error (https://github.com/hickory-dns/hickory-dns/pull/2224). Waiting for approval @djc @bluejekyll

File: .github/CODEOWNERS
Comment (bluejekyll) [inline .github/CODEOWNERS:8]: Looks good.

File: .github/workflows/conformance.yml
Comment (bluejekyll) [reply .github/workflows/conformance.yml:12]: what's the length of time it takes today?

File: conformance/.github/workflows/ci.yml
Comment (bluejekyll) [inline conformance/.github/workflows/ci.yml:9]: any reason to pin this to a version like this?

File: conformance/Cargo.toml
Comment (bluejekyll) [inline conformance/Cargo.toml:1]: Do we need to ignore this in the parent workspace?

**Review by bluejekyll — COMMENTED** [review]
I'm not going to have time to go through all the code here, but the structure overall looks good. I think there are some tests we can probably move from hickory itself and into here as well. We can evaluate that later.

File: .github/CODEOWNERS
Comment (pvdrz) [reply .github/CODEOWNERS:8]: The last run took 3:34 minutes, but it can range from 2 to 5

File: .github/workflows/conformance.yml
Comment (pvdrz) [reply .github/workflows/conformance.yml:12]: The last run took 3:34 minutes, but it can range from 2 to 5

File: conformance/Cargo.toml
Comment (pvdrz) [reply conformance/Cargo.toml:1]: it shouldn't be an issue as the top level `Cargo.toml` does not contain `conformance` as a member.

File: .github/CODEOWNERS
Comment (bluejekyll) [reply .github/CODEOWNERS:2]: I haven't done that before, I'd be happy to configure that.

**Comment by bluejekyll** [comment]
looks like this is clean from a testing perspective. @djc, do you have any concerns?

**Review by bluejekyll — APPROVED** [review]
Can you just answer the comment about the pinned version?

Otherwise this is good.

File: conformance/.github/workflows/ci.yml
Comment (pvdrz) [reply conformance/.github/workflows/ci.yml:9]: It is done to avoid bothering contributors with fixing broken tests they are unfamiliar with as this conformance test suite is a new addition.

Comment (japaric) [reply conformance/.github/workflows/ci.yml:9]: > any reason to pin this to a version like this?

this whole directory should have been removed but I forgot to do so. it contains the CI settings used in the dnssec-tests repo. we used a pinned version of hickory-dns because we didn't have its source as part of the repository. this workflow file is not read by the GHA worker.

the new CI workflow, in `/.github/workflows/conformance.yml`, uses the latest version of hickory-dns which is what's included in the checkout of the repository.

File: conformance/packages/dns-test/Cargo.toml
Comment (japaric) [reply conformance/packages/dns-test/Cargo.toml:9]: given that we are using latest stable (1.78) instead of this project's MSRV (1.67) to run the conformance test suite, we can indeed use `std::OnceLock` (1.70) instead of `lazy_static!`. that'd be a good refactor to do after this PR is merged

Comment (bluejekyll) [reply conformance/packages/dns-test/Cargo.toml:9]: I'm not sure we "need" to change this, but, 1.70 was released a year ago at this point, I don't have any problem with us upgrading that, I think we really only have a mildly enforced goal of about 6 months or -3 releases.

**Comment by japaric** [comment]
@djc @bluejekyll if you are happy with the current state of this PR, could we land it soon-ish? we have some leftover PRs in the dnssec-tests repo that we'd like to re-submit against this repo

**Comment by djc** [comment]
Happy to merge this if you can rebase it to get rid of conflicts.

**Comment by japaric** [comment]
@djc I don't see merge conflicts between this PR and main. did you mean something else by "conflicts"?
2024-06-07T10:45:49+02:00

**Comment by djc** [comment]
Screenshot 2024-06-07 at 11 00 52

**Comment by japaric** [comment]
@djc rebased. let me know if the status still didn't change on your side

**Comment by djc** [comment]
Status still showing the same thing... Very weird.

**Comment by japaric** [comment]
I wonder what kind of merge strategy GitHub is trying to use. these two commands work locally for me
``` console
$ git checkout main

$ # regular merge with merge commit
$ git merge --no-ff -m 'merge PR' ja-import-conformance-tests

$ # rebase and merge (no merge commit)
$ git merge --ff-only ja-import-conformance-tests

**Comment by bluejekyll** [comment]
Thanks for reviewing @djc!

**Pull request merged**
